From 38f809a13fae06f8dd3952681c6aa026dae5e47b Mon Sep 17 00:00:00 2001
From: Karmjit Mahil <karmjit.mahil@igalia.com>
Date: Wed, 14 Oct 2026 09:55:53 +0100
Subject: [PATCH 1/2] tu: Skip zero sized inline uniform blocks

This reverts commit 7bdcbe11ac507b0b96d2333aaeab5cf63de9ad10.

This isn't a workaround. See:
- VkDescriptorSetLayoutBinding:
  "If descriptorCount is zero this binding entry is reserved and the
  resource must not be accessed from any stage via this binding within
  any pipeline using the set layout."
- "maxPerStageDescriptorInlineUniformBlocks is the maximum number of
  inline uniform block bindings that can be accessible to a single
  shader stage in a pipeline layout."
- VkWriteDescriptorSet: "Consecutive Binding Updates" section.

Fix for: dEQP-VK.binding_model.inline_uniform_blocks.consecutive_copy_descriptor_count_0

Signed-off-by: Karmjit Mahil <karmjit.mahil@igalia.com>
---
 src/freedreno/vulkan/tu_shader.cc | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/freedreno/vulkan/tu_shader.cc b/src/freedreno/vulkan/tu_shader.cc
index 0c79d5db..4d9f9158 100644
--- a/src/freedreno/vulkan/tu_shader.cc
+++ b/src/freedreno/vulkan/tu_shader.cc
@@ -1566,6 +1566,9 @@ tu_lower_io(nir_shader *shader, struct tu_device *dev,
                mesa_to_vk_shader_stage(shader->info.stage)))
             continue;
 
+         if (binding->size == 0)
+            continue;
+
          /* If we don't know the size at compile time due to a variable
           * descriptor count, then with descriptor buffers we cannot know
           * how much space the real inline uniform has. In this case we fall

From 56bd21d18bb211aafb4b0423f479eb8c94c4c914 Mon Sep 17 00:00:00 2001
From: Karmjit Mahil <karmjit.mahil@igalia.com>
Date: Wed, 14 Oct 2026 10:44:50 +0100
Subject: [PATCH 2/2] tu: Skip reserved bindings when updating inline uniform
 blocks

With "Consecutive Binding Updates" the update can span multiple
bindings. We have to skip bindings with 0 descriptors (i.e. reserved
bindings) which might be of a different type than INLINE_UNIFORM_BLOCK.

VUID-VkWriteDescriptorSet-descriptorCount-00317:
  "All consecutive bindings updated via a single VkWriteDescriptorSet
  structure, except those with a descriptorCount of zero, must have
  identical descriptorType"

A bunch of other VUID have: "except those with a descriptorCount of
zero, must have ..."

The "Consecutive Binding Updates" sections generally indicates:
  "If a binding has a descriptorCount of zero, it is skipped. This
  behavior applies recursively, with the update affecting consecutive
  bindings as needed to update all descriptorCount descriptors."

Fix for:
dEQP-VK.binding_model.inline_uniform_blocks.consecutive_write_chain_with_0_non_inline_bindings{,_update_with_template}

Signed-off-by: Karmjit Mahil <karmjit.mahil@igalia.com>
---
 src/freedreno/vulkan/tu_descriptor_set.cc | 21 ++++++++++++++-------
 1 file changed, 14 insertions(+), 7 deletions(-)

diff --git a/src/freedreno/vulkan/tu_descriptor_set.cc b/src/freedreno/vulkan/tu_descriptor_set.cc
index 18b9299a..226798ee 100644
--- a/src/freedreno/vulkan/tu_descriptor_set.cc
+++ b/src/freedreno/vulkan/tu_descriptor_set.cc
@@ -1441,7 +1441,8 @@ tu_update_descriptor_sets(const struct tu_device *device,
          uint32_t dst_offset = writeset->dstArrayElement;
          do {
             uint8_t *dst = reinterpret_cast<uint8_t *>(ptr) + dst_offset;
-            uint32_t binding_size = binding_layout->size - dst_offset;
+            bool is_reserved = binding_layout->array_size == 0;
+            uint32_t binding_size = is_reserved ? 0 : (binding_layout->size - dst_offset);
             uint32_t to_write = MIN2(remaining, binding_size);
             memcpy(dst, src, to_write);
 
@@ -1550,10 +1551,12 @@ tu_update_descriptor_sets(const struct tu_device *device,
          uint32_t dst_start = copyset->dstArrayElement;
          uint8_t *src = reinterpret_cast<uint8_t *>(src_ptr) + src_start;
          uint8_t *dst = reinterpret_cast<uint8_t *>(dst_ptr) + dst_start;
+         bool src_is_reserved = src_binding_layout->array_size == 0;
+         bool dst_is_reserved = dst_binding_layout->array_size == 0;
          uint32_t src_remaining =
-            src_binding_layout->size - src_start;
+            src_is_reserved ? 0 : (src_binding_layout->size - src_start);
          uint32_t dst_remaining =
-            dst_binding_layout->size - dst_start;
+            dst_is_reserved ? 0 : (dst_binding_layout->size - dst_start);
          do {
             uint32_t to_write = MIN3(remaining, src_remaining, dst_remaining);
             memcpy(dst, src, to_write);
@@ -1570,14 +1573,16 @@ tu_update_descriptor_sets(const struct tu_device *device,
                src_binding_layout++;
                src_ptr = src_set->mapped_ptr + src_binding_layout->offset / 4;
                src = reinterpret_cast<uint8_t *>(src_ptr);
-               src_remaining = src_binding_layout->size;
+               src_is_reserved = src_binding_layout->array_size == 0;
+               src_remaining = src_is_reserved ? 0 : src_binding_layout->size;
             }
 
             if (dst_remaining == 0) {
                dst_binding_layout++;
                dst_ptr = dst_set->mapped_ptr + dst_binding_layout->offset / 4;
                dst = reinterpret_cast<uint8_t *>(dst_ptr);
-               dst_remaining = dst_binding_layout->size;
+               dst_is_reserved = dst_binding_layout->array_size == 0;
+               dst_remaining = dst_is_reserved ? 0 : dst_binding_layout->size;
             }
          } while (remaining > 0);
 
@@ -1666,7 +1671,8 @@ tu_CreateDescriptorUpdateTemplate(
       uint32_t remaining = entry->descriptorCount;
       uint32_t dst_start = entry->dstArrayElement;
       do {
-         uint32_t size = binding_layout->size;
+         bool is_reserved = binding_layout->array_size == 0;
+         uint32_t size = is_reserved ? 0 : binding_layout->size;
          uint32_t count = MIN2(remaining, size - dst_start);
          remaining -= count;
          binding_layout++;
@@ -1722,7 +1728,8 @@ tu_CreateDescriptorUpdateTemplate(
          do {
             dst_offset =
                binding_layout->offset + dst_start;
-            uint32_t size = binding_layout->size;
+            bool is_reserved = binding_layout->array_size == 0;
+            uint32_t size = is_reserved ? 0 : binding_layout->size;
             uint32_t count = MIN2(remaining, size - dst_start);
             templ->entry[j++] = (struct tu_descriptor_update_template_entry) {
                .descriptor_type = entry->descriptorType,
